File: 1018.binary-prefix-divisible-by-5.cpp
/*
 * @lc app=leetcode id=1018 lang=cpp
 *
 * [1018] Binary Prefix Divisible By 5
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<bool> prefixesDivBy5(vector<int>& nums) {
    vector<bool> ret;
    int base_10 = 0;
    for (const int num : nums) {
      base_10 = ((base_10 << 1) | num) % 5;
      ret.push_back(base_10 == 0);
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> nums;
  vector<bool> exp;
  nums = {0, 1, 1};
  exp = {true, false, false};
  ASSERT_EQ(s.prefixesDivBy5(nums), exp);
  nums = {1, 1, 1};
  exp = {false, false, false};
  ASSERT_EQ(s.prefixesDivBy5(nums), exp);
  nums = {0, 1, 1, 1, 1, 1};
  exp = {true, false, false, false, true, false};
  ASSERT_EQ(s.prefixesDivBy5(nums), exp);
  nums = {1, 1, 1, 0, 1};
  exp = {false, false, false, false, false};
  ASSERT_EQ(s.prefixesDivBy5(nums), exp);
  return 0;
}

File: 1018.binary-prefix-divisible-by-5.rs
/*
 * @lc app=leetcode id=1018 lang=rust
 *
 * [1018] Binary Prefix Divisible By 5
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn prefixes_div_by5(nums: Vec<i32>) -> Vec<bool> {
        let mut ret = Vec::<bool>::new();
        let mut num = 0;
        for n in nums {
            num = (num * 2 + n) % 5;
            ret.push(num == 0);
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::prefixes_div_by5(vec![0, 1, 1]),
        vec![true, false, false]
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::prefixes_div_by5(vec![1, 1, 1]),
        vec![false, false, false]
    );
}
